File: app/src/main/java/dev/ragnarok/fenrir/model/Community.java
package dev.ragnarok.fenrir.model;

import static dev.ragnarok.fenrir.util.Utils.firstNonEmptyString;

import android.os.Parcel;

import dev.ragnarok.fenrir.CheckDonate;
import dev.ragnarok.fenrir.module.parcel.ParcelNative;
import dev.ragnarok.fenrir.util.Utils;


public class Community extends Owner {

    public static final Creator<Community> CREATOR = new Creator<Community>() {
        @Override
        public Community createFromParcel(Parcel in) {
            return new Community(in);
        }

        @Override
        public Community[] newArray(int size) {
            return new Community[size];
        }
    };
    public static final ParcelNative.Creator<Community> NativeCreator = Community::new;
    private final int id;
    private String name;
    private String screenName;
    private int closed;
    private boolean admin;
    private int adminLevel;
    private boolean member;
    private int membersCount;
    private int memberStatus;
    private int type;
    private String photo50;
    private String photo100;
    private String photo200;
    private boolean verified;

    public Community(int id) {
        super(OwnerType.COMMUNITY);
        this.id = id;
    }

    protected Community(Parcel in) {
        super(in);
        id = in.readInt();
        name = in.readString();
        screenName = in.readString();
        closed = in.readInt();
        admin = in.readByte() != 0;
        adminLevel = in.readInt();
        member = in.readByte() != 0;
        membersCount = in.readInt();
        memberStatus = in.readInt();
        type = in.readInt();
        photo50 = in.readString();
        photo100 = in.readString();
        photo200 = in.readString();
        verified = in.readByte() != 0;
    }

    protected Community(ParcelNative in) {
        super(in);
        id = in.readInt();
        name = in.readString();
        screenName = in.readString();
        closed = in.readInt();
        admin = in.readByte() != 0;
        adminLevel = in.readInt();
        member = in.readByte() != 0;
        membersCount = in.readInt();
        memberStatus = in.readInt();
        type = in.readInt();
        photo50 = in.readString();
        photo100 = in.readString();
        photo200 = in.readString();
        verified = in.readByte() != 0;
    }

    @Override
    public String getFullName() {
        return name;
    }

    @Override
    public int getOwnerId() {
        return -Math.abs(id);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        super.writeToParcel(dest, flags);
        dest.writeInt(id);
        dest.writeString(name);
        dest.writeString(screenName);
        dest.writeInt(closed);
        dest.writeByte((byte) (admin ? 1 : 0));
        dest.writeInt(adminLevel);
        dest.writeByte((byte) (member ? 1 : 0));
        dest.writeInt(membersCount);
        dest.writeInt(memberStatus);
        dest.writeInt(type);
        dest.writeString(photo50);
        dest.writeString(photo100);
        dest.writeString(photo200);
        dest.writeByte((byte) (verified ? 1 : 0));
    }

    @Override
    public void writeToParcelNative(ParcelNative dest) {
        super.writeToParcelNative(dest);
        dest.writeInt(id);
        dest.writeString(name);
        dest.writeString(screenName);
        dest.writeInt(closed);
        dest.writeByte((byte) (admin ? 1 : 0));
        dest.writeInt(adminLevel);
        dest.writeByte((byte) (member ? 1 : 0));
        dest.writeInt(membersCount);
        dest.writeInt(memberStatus);
        dest.writeInt(type);
        dest.writeString(photo50);
        dest.writeString(photo100);
        dest.writeString(photo200);
        dest.writeByte((byte) (verified ? 1 : 0));
    }

    public String getName() {
        return name;
    }

    public Community setName(String name) {
        this.name = name;
        return this;
    }

    public String getScreenName() {
        return screenName;
    }

    public Community setScreenName(String screenName) {
        this.screenName = screenName;
        return this;
    }

    public int getClosed() {
        return closed;
    }

    public Community setClosed(int closed) {
        this.closed = closed;
        return this;
    }

    public boolean isAdmin() {
        return admin;
    }

    public Community setAdmin(boolean admin) {
        this.admin = admin;
        return this;
    }

    public int getAdminLevel() {
        return adminLevel;
    }

    public Community setAdminLevel(int adminLevel) {
        this.adminLevel = adminLevel;
        return this;
    }

    public boolean isMember() {
        return member;
    }

    public Community setMember(boolean member) {
        this.member = member;
        return this;
    }

    public int getMembersCount() {
        return membersCount;
    }

    public Community setMembersCount(int membersCount) {
        this.membersCount = membersCount;
        return this;
    }

    public int getMemberStatus() {
        return memberStatus;
    }

    public Community setMemberStatus(int memberStatus) {
        this.memberStatus = memberStatus;
        return this;
    }

    public int getType() {
        return type;
    }

    public Community setType(int type) {
        this.type = type;
        return this;
    }

    public String getPhoto50() {
        return photo50;
    }

    public Community setPhoto50(String photo50) {
        this.photo50 = photo50;
        return this;
    }

    public String getPhoto100() {
        return photo100;
    }

    public Community setPhoto100(String photo100) {
        this.photo100 = photo100;
        return this;
    }

    public String getPhoto200() {
        return photo200;
    }

    public Community setPhoto200(String photo200) {
        this.photo200 = photo200;
        return this;
    }

    @Override
    public String get100photoOrSmaller() {
        return firstNonEmptyString(photo100, photo50);
    }

    @Override
    public String getDomain() {
        return screenName;
    }

    @Override
    public String getMaxSquareAvatar() {
        return firstNonEmptyString(photo200, photo100, photo50);
    }

    @Override
    public String getOriginalAvatar() {
        return getMaxSquareAvatar();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public int getId() {
        return id;
    }

    @Override
    public boolean isDonated() {
        return Utils.isValueAssigned(getOwnerId(), CheckDonate.donatedOwnersLocal);
    }

    @Override
    public boolean isVerified() {
        return verified || isDonated();
    }

    public Community setVerified(boolean verified) {
        this.verified = verified;
        return this;
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/Community.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.Keep
import dev.ragnarok.fenrir.CheckDonate
import dev.ragnarok.fenrir.getBoolean
import dev.ragnarok.fenrir.module.parcel.ParcelNative
import dev.ragnarok.fenrir.putBoolean
import dev.ragnarok.fenrir.util.Utils.firstNonEmptyString
import dev.ragnarok.fenrir.util.Utils.isValueAssigned
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.math.abs

@Keep
@Serializable
@SerialName("community")
class Community : Owner {
    val id: Int
    override var fullName: String? = null
        private set
    override val domain: String?
        get() = screenName
    private var screenName: String? = null
    var closed = 0
        private set
    var isAdmin = false
        private set
    var adminLevel = 0
        private set
    var isMember = false
        private set
    var membersCount = 0
        private set
    var memberStatus = 0
        private set
    var communityType = 0
        private set
    var photo50: String? = null
        private set
    var photo100: String? = null
        private set
    var photo200: String? = null
        private set
    var verified = false
        private set
    var isBlacklisted = false
        private set
    var hasUnseenStories = false
        private set

    constructor(id: Int) : super(OwnerType.COMMUNITY) {
        this.id = id
    }

    internal constructor(parcel: Parcel) : super(parcel) {
        id = parcel.readInt()
        fullName = parcel.readString()
        screenName = parcel.readString()
        closed = parcel.readInt()
        isAdmin = parcel.getBoolean()
        adminLevel = parcel.readInt()
        isMember = parcel.getBoolean()
        membersCount = parcel.readInt()
        memberStatus = parcel.readInt()
        communityType = parcel.readInt()
        photo50 = parcel.readString()
        photo100 = parcel.readString()
        photo200 = parcel.readString()
        verified = parcel.getBoolean()
        isBlacklisted = parcel.getBoolean()
        hasUnseenStories = parcel.getBoolean()
    }

    internal constructor(parcel: ParcelNative) : super(parcel) {
        id = parcel.readInt()
        fullName = parcel.readString()
        screenName = parcel.readString()
        closed = parcel.readInt()
        isAdmin = parcel.readBoolean()
        adminLevel = parcel.readInt()
        isMember = parcel.readBoolean()
        membersCount = parcel.readInt()
        memberStatus = parcel.readInt()
        communityType = parcel.readInt()
        photo50 = parcel.readString()
        photo100 = parcel.readString()
        photo200 = parcel.readString()
        verified = parcel.readBoolean()
        isBlacklisted = parcel.readBoolean()
        hasUnseenStories = parcel.readBoolean()
    }

    override val ownerId: Int
        get() = -abs(id)

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        super.writeToParcel(parcel, flags)
        parcel.writeInt(id)
        parcel.writeString(fullName)
        parcel.writeString(screenName)
        parcel.writeInt(closed)
        parcel.putBoolean(isAdmin)
        parcel.writeInt(adminLevel)
        parcel.putBoolean(isMember)
        parcel.writeInt(membersCount)
        parcel.writeInt(memberStatus)
        parcel.writeInt(communityType)
        parcel.writeString(photo50)
        parcel.writeString(photo100)
        parcel.writeString(photo200)
        parcel.putBoolean(verified)
        parcel.putBoolean(isBlacklisted)
        parcel.putBoolean(hasUnseenStories)
    }

    override fun writeToParcelNative(dest: ParcelNative) {
        super.writeToParcelNative(dest)
        dest.writeInt(id)
        dest.writeString(fullName)
        dest.writeString(screenName)
        dest.writeInt(closed)
        dest.writeBoolean(isAdmin)
        dest.writeInt(adminLevel)
        dest.writeBoolean(isMember)
        dest.writeInt(membersCount)
        dest.writeInt(memberStatus)
        dest.writeInt(communityType)
        dest.writeString(photo50)
        dest.writeString(photo100)
        dest.writeString(photo200)
        dest.writeBoolean(verified)
        dest.writeBoolean(isBlacklisted)
        dest.writeBoolean(hasUnseenStories)
    }

    @AbsModelType
    override fun getModelType(): Int {
        return AbsModelType.MODEL_COMMUNITY
    }

    fun setName(name: String?): Community {
        fullName = name
        return this
    }

    fun setScreenName(screenName: String?): Community {
        this.screenName = screenName
        return this
    }

    fun setHasUnseenStories(hasUnseenStories: Boolean): Community {
        this.hasUnseenStories = hasUnseenStories
        return this
    }

    fun setBlacklisted(isBlacklisted: Boolean): Community {
        this.isBlacklisted = isBlacklisted
        return this
    }

    fun setClosed(closed: Int): Community {
        this.closed = closed
        return this
    }

    fun setAdmin(admin: Boolean): Community {
        isAdmin = admin
        return this
    }

    fun setAdminLevel(adminLevel: Int): Community {
        this.adminLevel = adminLevel
        return this
    }

    fun setMember(member: Boolean): Community {
        isMember = member
        return this
    }

    fun setMembersCount(membersCount: Int): Community {
        this.membersCount = membersCount
        return this
    }

    fun setMemberStatus(memberStatus: Int): Community {
        this.memberStatus = memberStatus
        return this
    }

    fun setCommunityType(type: Int): Community {
        this.communityType = type
        return this
    }

    fun setPhoto50(photo50: String?): Community {
        this.photo50 = photo50
        return this
    }

    fun setPhoto100(photo100: String?): Community {
        this.photo100 = photo100
        return this
    }

    fun setPhoto200(photo200: String?): Community {
        this.photo200 = photo200
        return this
    }

    override fun get100photoOrSmaller(): String? {
        return firstNonEmptyString(photo100, photo50)
    }

    override val maxSquareAvatar: String?
        get() = firstNonEmptyString(photo200, photo100, photo50)
    override val originalAvatar: String?
        get() = maxSquareAvatar

    override fun describeContents(): Int {
        return 0
    }

    override val isDonated: Boolean
        get() = isValueAssigned(ownerId, CheckDonate.donatedOwnersLocal)

    override val isVerified: Boolean
        get() {
            return verified || isDonated
        }

    override val isHasUnseenStories: Boolean
        get() {
            return hasUnseenStories
        }

    fun setVerified(verified: Boolean): Community {
        this.verified = verified
        return this
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Community> = object : Parcelable.Creator<Community> {
            override fun createFromParcel(parcel: Parcel): Community {
                return Community(parcel)
            }

            override fun newArray(size: Int): Array<Community?> {
                return arrayOfNulls(size)
            }
        }
        val NativeCreator: ParcelNative.Creator<Community> =
            object : ParcelNative.Creator<Community> {
                override fun readFromParcelNative(dest: ParcelNative): Community {
                    return Community(dest)
                }

            }
    }
}